RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1T.cc

```diff
@@ -124,246 +165,753 @@ void L2MuonSeedGeneratorFromL1T::produce(edm::Event& iEvent, const edm::EventSet
+                          TrajectorySeed::const_iterator
+                          tsci  = selOffseeds[theL1][theOffs]->recHits().first,
+                          tscie = selOffseeds[theL1][theOffs]->recHits().second;
+                          for(; tsci!=tscie; ++tsci) {
+                            Newcontainer.push_back(*tsci);
+                          }
+                          output->push_back(L2MuonTrajectorySeed(seedTSOS,Newcontainer,alongMomentum,
+                                            MuonRef(muColl,  distance(muColl->begin(muColl->getFirstBX()),Newit)  )));
+          }
+        }
+      }
+
+      else if(matchType==2) {
+
+        for(nL1=0; nL1 < NmuColl; ++nL1) {
+          double tempDR = 999;
+          int theL1 = 0;
+          int theOffs = 0;
+
+          for(i=0; i<NmuColl; ++i) {
+            for(j=0; j<NofflineSeed1; ++j) {
+              if(tempDR > dRmtx[i][j]) {
+                tempDR = dRmtx[i][j];
+                theL1 = i;
+                theOffs = j;
+              }
+            }
+          }
+
+          auto Newit = muColl->begin(muColl->getFirstBX()) + theL1;
+
+          double NewdRcone = matchingDR[0];
+          if ( fabs(Newit->eta()) < etaBins.back() ){
+            std::vector<double>::iterator lowEdge = std::upper_bound (etaBins.begin(), etaBins.end(), fabs(Newit->eta()));
+            NewdRcone    =  matchingDR.at( lowEdge - etaBins.begin() - 1);
+          }
+
+          if( !(tempDR < NewdRcone) )  continue;
+
+          // Remove row and column for given (L1mu, offSeed)
+          for(i=0; i<NmuColl; ++i) {
+            dRmtx[i][theOffs] = 999;
+          }
+          for(j=0; j<NofflineSeed1; ++j) {
+            dRmtx[theL1][j] = 999;
+          }
+
+          if( selOffseeds[theL1][theOffs] != nullptr ) {
+            //put given L1mu and offSeed to output
+            edm::OwnVector<TrackingRecHit> Newcontainer;
+
+            PTrajectoryStateOnDet const & seedTSOS = selOffseeds[theL1][theOffs] ->startingState();
+                          TrajectorySeed::const_iterator
+                          tsci  = selOffseeds[theL1][theOffs]->recHits().first,
+                          tscie = selOffseeds[theL1][theOffs]->recHits().second;
+                          for(; tsci!=tscie; ++tsci) {
+                            Newcontainer.push_back(*tsci);
+                          }
+                          output->push_back(L2MuonTrajectorySeed(seedTSOS,Newcontainer,alongMomentum,
+                                            MuonRef(muColl,  distance(muColl->begin(muColl->getFirstBX()),Newit)  )));
+          }
+        }
+      }
+
+      else if(matchType==3) {
+
+        for(nL1=0; nL1 < NmuColl; ++nL1) {
+          double tempDR = 999;
+          int theL1 = 0;
+          int theOffs = 0;
+          auto theit = muColl->begin(muColl->getFirstBX());
+
+          // L1Q > 10 (L1Q = 12)
+          for(i=0; i<NmuColl; ++i) {
```

Comment:
Instead of making these three nested loops on nMuColl, you can just compare the quality: if it is above or equal the highest one encountered so far in the loop, you make the loop on the offline seeds, check tempDR, etc
(If I understand correctly how it works...)